src/metabase/upload.clj

```diff
@@ -340,3 +348,130 @@
       (catch Throwable e
         (driver/drop-table! driver db-id table-name)
         (throw (ex-info (ex-message e) {:status-code 400}))))))
+
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                                 public interface                                               |
+;;; +----------------------------------------------------------------------------------------------------------------+
+
+(defn- scan-and-sync-table!
+  [database table]
+  (sync-fields/sync-fields-for-table! database table)
+  (future
+    (sync/sync-table! table)))
```

Comment:
I find this sync code a bit strange:
1. `sync-table!` will trigger `sync-fields-for-table!` too, so why do we need to do it separately?
2. why `sync-table!` is in a future but `sync-fields-for-table!` is not?

Reply:
Indeed it's not perfect, and we're doing redundant work. But when we wrote this code I assume we didn't want to spend time expanding the sync interface for this use case. It feels a little risky to call individual functions inside the sync namespaces that are not tested individually and are implementation details of more "public" sync functions such as `sync/sync-table!`.

As for 2, we need the `sync-fields-for-table!` to run synchronously, but we don't want to wait for the table to be analyzed before continuing with the rest of the steps in `upload-csv!`.